services/galley/src/Galley/API/Update.hs

```diff
@@ -105,13 +113,114 @@ unblockConv (usr ::: conn ::: cnv) = do
     conv' <- acceptOne2One usr conv conn
     setStatus status200 . json <$> conversationView usr conv'
 
-joinConversation :: UserId ::: ConnId ::: ConvId ::: JSON -> Galley Response
-joinConversation (zusr ::: zcon ::: cnv ::: _) = do
+updateConversationAccess :: UserId ::: ConnId ::: ConvId ::: Request ::: JSON -> Galley Response
+updateConversationAccess (usr ::: zcon ::: cnv ::: req ::: _ ) = do
+    body <- fromBody req invalidPayload
+    let targetAccess = Set.fromList (toList (cupAccess body))
+    -- checks and balances
+    when (PrivateAccess `elem` targetAccess) $
+        throwM invalidTargetAccess
+    (bots, users) <- botsAndUsers <$> Data.members cnv
+    unless (usr `isMember` users) $
+        throwM convNotFound
+    conv <- Data.conversation cnv >>= ifNothing convNotFound
+    let currentAccess = Set.fromList (toList $ Data.convAccess conv)
+    case Data.convType conv of
+        SelfConv    -> throwM invalidSelfOp
+        One2OneConv -> throwM invalidOne2OneOp
+        ConnectConv -> throwM invalidConnectOp
+        _           -> return ()
+    -- remove conversation codes if CodeAccess is revoked
+    when (CodeAccess `elem` currentAccess && CodeAccess `notElem` targetAccess) $ do
+        key <- mkKey cnv
+        Data.deleteCode key
+    -- special case TeamConversation
+    case Data.convTeam conv of
+        Nothing     ->
+            when (TeamAccess `elem` targetAccess) $
+                throwM $ invalidTargetAccess
+        Just tid    -> do
+            tMembers <- Data.teamMembers tid
+            -- only team members can change access mode
+            unless (usr `elem` (memId <$> users)) $
+                throwM $ accessDenied
+            -- remove non-team users if the only mode left is TeamAccess
+            when (targetAccess == Set.fromList [TeamAccess]) $
+                handleTeamOnly tid tMembers users bots conv
+    -- update cassandra & send event
+    now <- liftIO getCurrentTime
+    let e = Event ConvAccessUpdate cnv usr now (Just $ EdConvAccessUpdate body )
+    unless (currentAccess == targetAccess) $ do
```

Comment:
Maybe we can check this in the very beginning and make the whole handler a no-op?